### DISCUSSION TITLE
Question about how the keyboard listener handles RTs

### DISCUSSION THREAD

**Discussion by sp-conway** [jspsych/jsPsych#2174] [Q&A] [answered]
Hi,
I ended up having to write my own (sloppy, I'm sure) plugin for an experimental task (essentially, a perceptual categorization experiment in which participants have to pick the best stimulus from either 1, 2, or 3 vertical lines, drawn via canvas). I wanted to make sure the fixation cross was always located in the same position as the line(s), so a fixation cross was implemented in the plugin too. That is, each trial goes fixation cross - clear screen - draw stimuli/ask for response. All specified in my plugin.

Essentially, I use the `jsPsych.pluginAPI.setTimeout` method to specify how long to display the fixation cross and how long to show a blank screen between cross and stimuli. 

What I realized I need to do is prevent the keyboard listener from starting until AFTER the stimuli are drawn (so participants can't skip through all trials during the fixation cross). I am trying to do so below, but I'm confused about how the keyboard listener works (specifically, what rts will be timelocked to). 

`fix_dur` is a variable specifying duration of my fixation cross, `post_fix_dur` is a variable specifying the duration of a blank screen following the fixation cross.

`draw_fix_cross()` draws a fixation cross, `clear_screen()` clears the screen. `drawline1()`, `drawline2()`, and `drawline3()` draw 1, 2, and 3 stimuli, respectively (the details of these functions are unimportant to my issue).

Basically, can anyone provide any insight into how they keyboard listener works, and how to workout my timings such that I can trust that my rts are measured from stimulus onset?

```
jsPsych.pluginAPI.setTimeout(function(){
      var keyboardListener = jsPsych.pluginAPI.getKeyboardResponse({
           callback_function: after_response,
           valid_responses: trial.choices,
           rt_method: 'performance',
           persist: false,
           allow_held_key: false
      });
    }, fix_dur); // if I do fix_dur this is too short, and fix_dur+post_fix_dur appears to be too long?

// Single stimulus presentation 
if(n_stim==1){
  draw_fix_cross();
  jsPsych.pluginAPI.setTimeout(clear_screen, fix_dur);
  jsPsych.pluginAPI.setTimeout(drawline1, post_fix_dur);
// multiple stimuli presentation (2 or 3)
} else if (n_stim==2) {
  draw_fix_cross();
  jsPsych.pluginAPI.setTimeout(clear_screen, fix_dur);
  jsPsych.pluginAPI.setTimeout(drawline2, post_fix_dur);
} else if (n_stim==3){
  draw_fix_cross();
  jsPsych.pluginAPI.setTimeout(clear_screen, fix_dur);
  jsPsych.pluginAPI.setTimeout(drawline3, post_fix_dur);
}
```

**Comment by jodeleeuw** [comment, marked as answer]
Hi @sp-conway,

The RT is measured from the moment that the keyboard listener is created by calling `getKeyboardResponse`. 

What I would recommend is rather than trying to schedule all of the drawing events with different calls to `setTimeout` you should chain together these events so that one triggers the next. For example:

```js
function begin(){
	draw_fixation_cross();
	jsPsych.pluginAPI.setTimeout(blank_screen, fix_dur);
}

function blank_screen(){
	clear_screen();
	jsPsych.pluginAPI.setTimeout(draw_lines, post_fix_dur);
}

function draw_lines(){
	if(n_stim == 1){
		drawline1();
	}
	if(n_stim == 2){
		drawline2();
	}
	if(n_stim == 3){
		drawline3();
	}
	jsPsych.pluginAPI.getKeyboardResponse({ ... })
}
```

I think what was happening in your code is that you were setting up three timers with the same start time. Let's say that:

```js
var fix_dur = 500;
var post_fix_dur = 1000;
```

Then when you set this up:

```js
draw_fix_cross();
jsPsych.pluginAPI.setTimeout(clear_screen, fix_dur);
jsPsych.pluginAPI.setTimeout(drawline1, post_fix_dur);
```

The `clear_screen` function will trigger approx 500 ms after `draw_fixation_cross` and `drawline1` will trigger approx 1000 ms **after `draw_fixation_cross`**. I'm guessing that you imagined that the second `setTimeout` wasn't started until the first one completes, but they are actually initialized at basically the same time.

**Reply by sp-conway** [reply]
Thanks so much @jodeleeuw ! This is extremely helpful. It's clearly better programming than what I had set up, and it clears up much of the confusion re: `jsPsych.pluginAPI.setTimeout`.

Would it make sense to set up the `draw_lines()` function like this? 
```
function draw_lines(){
        	if(n_stim == 1){
        		draw_line_1();
        	}
        	if(n_stim == 2){
        		draw_line_2();
        	}
        	if(n_stim == 3){
        		draw_line_3();
        	}
          var keyboard_listener = jsPsych.pluginAPI.getKeyboardResponse({
           callback_function: after_response,
           valid_responses: trial.choices,
           rt_method: 'performance',
           persist: false,
           allow_held_key: falseI
          });
          return keyboard_listener
        }
```

**Reply by jodeleeuw** [reply]
You don't really need to return the keyboard listener handler unless you are planning to do something with it later on, like cancel it using cancelKeyboardResponse. The main thing it will do is invoke the after_response function once a valid key event is registered. But it also doesn't really hurt to return it.
